Feature/media upload audio files

marecar3 opened the pull request:
Part of [#4842](https://github.com/wordpress-mobile/WordPress-Android/issues/4842)

This PR adds ability to upload audio files in Media.
Also it demanded changes in FluxC project which is described in this [FluxC PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/837). For testing purpose this PR has FluxC hash code of temp branch so that it can be build against FluxC changes.

Testing instructions 

Visually tutorials look like this:

**1. Navigate to My site**
![device-2018-07-12-012833](https://user-images.githubusercontent.com/28219092/42605643-7fe377e6-8579-11e8-877d-5593c5151a15.png)

**2. Open Media section**
![device-2018-07-12-012901](https://user-images.githubusercontent.com/28219092/42605664-95179cd2-8579-11e8-9b66-ee75961bd25b.png)

**3. Click on the "+"  button in top right corner**
![device-2018-07-12-012913](https://user-images.githubusercontent.com/28219092/42605682-ac90463e-8579-11e8-953d-24ddaeb742bf.png)

**4. Click on "Choose audio from device"**
![device-2018-07-12-012947](https://user-images.githubusercontent.com/28219092/42605703-c683b508-8579-11e8-8d79-f1d7b4aa0d72.png)

**5. Choose some audio file from your device**
![device-2018-07-12-012958](https://user-images.githubusercontent.com/28219092/42605712-d4d8d912-8579-11e8-832e-084e4630d712.png)

**6. With a bit of luck upload will pass with success**
![device-2018-07-12-013024](https://user-images.githubusercontent.com/28219092/42605783-27db9ae6-857a-11e8-94bc-4a2b1eb72b7e.png)

marecar3 commented on build.gradle:
```diff
@@ -69,5 +69,5 @@ subprojects {
 
 ext {
-    fluxCVersion = '51fee5847ca60895772eb5e5e4ca64a275cac543'
+    fluxCVersion = '06a11f4fee4eeda8b7d4b778684b778e367f4b7a'
```
This shouldn't end up in develop branch!
It's here in a case if someone want's to build the branch and to test this feature.

mzorz commented:
Hi @marecar3 ! Some comments about PR format:

> Part of #4842

Should we expect more PRs to come? otherwise I think you can say "Fixes #4842".

- Remember to also make a reference to the FluxC PR in the description above (bonus points if it's in **bold**) - you can follow this as an example: https://github.com/wordpress-mobile/WordPress-Android/pull/7949

- Another thing, can you add to the testing instructions whether we need any kind of special plan to test this, or should we be able to upload audio files to any site (be it .org or .com)? Also, do we need to setup or configure anything on the server?

marecar3 commented:
Hi @mzorz ,
Thanks for the comments. 

> Should we expect more PRs to come? otherwise I think you can say "Fixes #4842".

Yes, there will be another PR which will add ability to upload audio files from Editor.

> Remember to also make a reference to the FluxC PR in the description above (bonus points if it's in bold) - you can follow this as an example: #7949

Done with bold style.

> Another thing, can you add to the testing instructions whether we need any kind of special plan to test this, or should we be able to upload audio files to any site (be it .org or .com)? Also, do we need to setup or configure anything on the server?

Will do this when I get some extra time today.

nbradbury commented on WordPress/src/main/java/org/wordpress/android/util/WPMediaUtils.java:
```diff
@@ -264,6 +266,24 @@ private static Intent preparePictureLibraryIntent(String title, boolean multiSel
+    private static Intent prepareAudioLibraryIntent(String title, boolean multiSelect) {
+        Intent intent = new Intent(Intent.ACTION_GET_CONTENT);
+        intent.setType(MIME_DATA_TYPE_AUDIO);
+        if (multiSelect) {
```
Minor nit: this can be shortened to `if (multiSelect && Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2)`

marecar3 replied:
Yup. Fixed.

nbradbury commented:
@marecar3 This looks good to me. I went ahead and tested with a .org site (both with and without Jetpack enabled) and it worked just fine.

A couple of recommendations:

- Instead of targeting `develop` create a separate feature branch from `develop`, call it something like `feature/upload-audio`, then target this PR to that one. This way when you work on uploading from the editor, you can target that same feature branch and then when both PRs are merged into the feature branch we can talk about merging all at once into `develop`.
- If you have access to a Samsung device, try running this on it. I'm not sure if this is still the case, but I think at one time Samsung had their own picker which sometimes behaved differently than the stock one.

marecar3 commented:
@nbradbury Thanks for recommendations.

> Instead of targeting develop create a separate feature branch from develop, call it something like feature/upload-audio, then target this PR to that one. This way when you work on uploading from the editor, you can target that same feature branch and then when both PRs are merged into the feature branch we can talk about merging all at once into develop.

Good one! Will do this :)

> If you have access to a Samsung device, try running this on it. I'm not sure if this is still the case, but I think at one time Samsung had their own picker which sometimes behaved differently than the stock one.

I tested on Samsung S6 Edge and all went well except uploading of "wav" audio files. Android API on Samsung device returns mime type "audio/vnd.wave" for "wav" extension, so I update my PR in FLuxC project : https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/837/commits/2dbfeaab53bf670ec1a689449fdc07645714db7e for that special case.

loremattei commented:
Hey! I'm moving the milestone to 10.6 because 10.5 has been cut. Please, feel free to move it back to 10.5 if it has to land there: update it to target the release branch and ping me to update the build!

mzorz commented:
Nice job @marecar3! 

I've been checking the branch with different files and a regular free site, one thing that happened with these two files is that I couldn't upload them but still they were eligible for uploading as per the file picker:
https://www.dropbox.com/sh/rr7b7hry93vqq6v/AADWX-EuLMNL2L4e_5IsSNYIa?dl=0
(I can provide logcat logs to show exactly what happened but thought it was better to provide the files so you could test on your side if needed).
One of them failed right away, but the other one failed only after being completely uploaded, which is of course not desirable given transfers like this could incur in network costs.

To get a reference, I tried to upload on `wp-admin` on the web to see if it was feasible, but it doesn't let you pick audio files if the site can't accept these files: when trying to upload in `wp-admin`, audio files (.wav and .mp3 in particular as per my attempt) are all greyed out.

So 2 observations:

1) Could it be related to the fact that audio/video files are only allowed for[ Personal/Premium/Business plan-enabled sites](https://en.support.wordpress.com/accepted-filetypes/)?

If that's the case maybe we should replicate the same behavior and only allow show the `Choose audio from device` option for sites that have either Personal, Premium or Business plans enabled. What do you think?

2) When testing a Personal/Premium/Business site, [this example file fails to upload](https://www.dropbox.com/s/2nchqv0ok2emt7s/Sendung%20von%20Mittwoch%2C%2018.07.2018%202230%20Uhr?dl=0). Could you check why and whether this is something that can be fixed, please? (tip if you can't download it from dropbox, rename it to something simple like test.mp3). Note: this is an actual file as saved by a Podcast app I use on my phone, so it appears filtered by audio type when attempting to include it in the WordPress app, hence that's why I'm asking, how we should deal with it.

marecar3 commented:
Hey @mzorz ,

Thanks for your feedback, I really appreciate it

> I've been checking the branch with different files and a regular free site, one thing that happened with these two files is that I couldn't upload them but still they were eligible for uploading as per the file picker:
https://www.dropbox.com/sh/rr7b7hry93vqq6v/AADWX-EuLMNL2L4e_5IsSNYIa?dl=0
(I can provide logcat logs to show exactly what happened but thought it was better to provide the files so you could test on your side if needed).
One of them failed right away, but the other one failed only after being completely uploaded, which is of course not desirable given transfers like this could incur in network costs.

There is an general problem in the app when you decide to pick the file from some 3rd part application. If you use e.g. Total commander you will be able to pick file type which could be different from type that you attended to pick. As we can't control 3rd part applications do only thing we can do is to check file extension/mime type of the file which is picked from picker.

> Could it be related to the fact that audio/video files are only allowed for Personal/Premium/Business plan-enabled sites?

For some reason users can upload videos from mobile apps even they don't have proper plan (on the web app it will be presented to them that they need to get plan in order to upload video/audio files)

mzorz commented:
> There is an general problem in the app when you decide to pick the file from some 3rd part application. If you use e.g. Total commander you will be able to pick file type which could be different from type that you attended to pick. As we can't control 3rd part applications the only thing that we can do is to check file extension/mime type of the file which is picked from picker.

Ah - you're right. Nice find and thanks for clarifying.

> > Could it be related to the fact that audio/video files are only allowed for Personal/Premium/Business plan-enabled sites?

> For some reason users can upload videos from mobile apps even they don't have proper plan (on the web app it will be presented to them that they need to get plan in order to upload video/audio files

You're right here as well 👍 

If possible, let's do as per your suggestion over DM in Slack, I'm copying here for completeness:

>  we can check the URI (URI can provide us extension and mime type) right after file is picked and if it's different from supported, we can reject it, what do you think ?

That seems like a good option, as it would prevent the file from being attempted to upload only to fail at the end. Also it's ok to treat it like that only for audio files for now.

Thanks for your answers, these were spot on!

marecar3 commented:
Hey @mzorz 

> That seems like a good option, as it would prevent the file from being attempted to upload only to fail at the end. Also it's ok to treat it like that only for audio files for now.

I have added validation of audio file format. Also I updated visually tutorials in this PR which present the case when user choose file format other then audio.

As I added strings in this PR, should we mark them for localization process or it will be done automatically when this PR merges into develop branch ?

mzorz commented on WordPress/src/main/res/values/strings.xml:
```diff
@@ -151,6 +152,7 @@
     <string name="media_error_exceeds_memory_limit">File too large to be uploaded on this site</string>
     <string name="media_error_internal_server_error">Upload error. Try changing Optimize Images in your app\'s settings</string>
     <string name="media_error_timeout">Server took too long to respond</string>
+    <string name="media_error_incorrect_audio_format">We\’re sorry, this type of audio file is not supported</string>
```
We should use the apostrophe character, like this: `We\'re sorry,[...]`

marecar3 replied:
Fixed.

mzorz commented:
> I have added validation of audio file format. Also I updated visually tutorials in this PR which present the case when user choose file format other then audio.

👍 Thanks, just checked commit b30c2d5 and that looks good.

> As I added strings in this PR, should we mark them for localization process or it will be done automatically when this PR merges into develop branch ?

Good question :) - there's no need for us to take further action, it'll be collected and sent to translators in the next release cycle. Sometimes when we have longer strings or have situations where more needs to be explained, we also get i18n people involved.

Nice work, I think we can get this one merged after [this minor comment](https://github.com/wordpress-mobile/WordPress-Android/pull/8027/files#r204269603) is addressed
